## Problem 14: Longest Collatz Sequence

The following iterative sequence is defined for the set of positive integers:

$n \to n/2$ ($n$ is even)

$n \to 3n + 1$ ($n$ is odd)

Using the rule above and starting with 13, we generate the following sequence:

$13 \to 40 \to 20 \to 10 \to 5 \to 16 \to 8 \to 4 \to 2 \to 1$.

It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

*NOTE*: Once the chain starts the terms are allowed to go above one million.

In [1]:
''' 
Finds the length of the Collatz Sequence for a number using recursion.

We cache the results to speed up computation, when a longer sequence becomes a smaller sequence we have already calcualted,
we can use the cached results rather than recalculating.

eg:
>>> collatz_sequence_length(13)
>>> 10
'''

from functools import lru_cache

@lru_cache(maxsize=1_000_000)
def collatz_sequence_length(number : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds the length of the Collatz Sequence for a number using recursion. Takes the input number and checks if it is 1
        (the sequence has ended). If not, then calculates the next term of the sequence and calls the function with that
        value. Then adds one to the sequence length.
        
        Base case: number = 1 : sequence_length = 1.
        Even case: number % 2 == 0 : number /= 2, sequence_length += 1.
        Odd case : number % 2 != 0 : number = (3 * number) + 1, sequence_length += 1.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number in the sequence.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : sequence_length : The length of the Collatz Sequence for that number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> collatz_sequence_length(13)
        >>> 10
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 1: raise ValueError('Please enter an integer > 0 for the number argument.')
    
    if number == 1:
        return 1
    
    # n -> n/2
    if number % 2 == 0:
        return collatz_sequence_length(int(number/2)) + 1
    
    # n -> 3n + 1
    return collatz_sequence_length((3 * number) + 1) + 1

In [2]:
''' 
Finds the longest Collatz Sequence for numbers under max_number.

Loops through the numbers from 1 to max_number and calcualtes the Collatz Sequence length for each.
Then compares the length to the current max length and saves the highest.

eg:
>>> longest_collatz_sequence(13)
>>> (9, 20)
'''

def longest_collatz_sequence(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds the longest Collatz Sequence for numbers under max_number.

        Loops through the numbers from 1 to max_number and calcualtes the Collatz Sequence length for each.
        Then compares the length to the current max length and saves the highest.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : The maximum starting number in the Collatz Sequences.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : The following information is returned:
            : int : Starting number of the longest Collatz Sequence.
            : int : Length of the longest Collatz Sequence.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> longest_collatz_sequence(13)
        >>> (9, 20)
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 1: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    longest_sequence_length = 0
    longest_sequence_number = 0
    
    for number in range(1, max_number):
        sequence_length = collatz_sequence_length(number)
        if sequence_length > longest_sequence_length:
            longest_sequence_length = sequence_length
            longest_sequence_number = number
    
    return longest_sequence_number, longest_sequence_length

In [3]:
problem_max_number = 1_000_000

solution = longest_collatz_sequence(problem_max_number)

print(f'The longest Collatz Sequence with starting number below {problem_max_number:,} starts with {solution[0]:,} and has a length of {solution[1]:,}.')

The longest Collatz Sequence with starting number below 1,000,000 starts with 837,799 and has a length of 525.


### Problem 14 Solution:

The longest Collatz Sequence with starting number below 1,000,000 starts with 837,799 and has a length of 525.